momentGW/thc.py

```diff
@@ -0,0 +1,270 @@
+            ZD_only[i] = np.einsum("PQ,PQ->PQ", Yea_max, YiP) + np.einsum("PQ,PQ->PQ", Yei_max, YaP)
+            ZD_temp = np.zeros((self.naux, self.naux))
+            for j in range(1, i):
+                Yei = np.einsum("i,iP,iQ->PQ", (-1) ** (j) * self.ei ** (j), self.XiP, self.XiP)
+                Yea = np.einsum("a,aP,aQ->PQ", binom(i, j) * self.ea ** (i - j), self.XaP, self.XaP)
+                ZD_only[i] += np.einsum("PQ,PQ->PQ", Yea, Yei)
+                if j == i - 1:
+                    Z_left += np.einsum("PQ,QR->PR", self.Z, ZD_only[j]) * 2
+                else:
+                    Z_left += (
+                        np.einsum("PQ,QR,RS->PS", self.Z, ZD_only[i - 1 - j], ZD_left[i - j]) * 2
+                    )
+                ZD_temp += np.einsum("PQ,QR->PR", ZD_only[j], ZD_left[j])
+            zeta[i] = ZD_only[i] + ZD_temp + np.einsum("PQ,QR->PR", self.Z_prime, Z_left)
+            cput1 = lib.logger.timer(self.gw, "Zeta %d" % i, *cput1)
+        return zeta
+
+    def build_THC_se_moments(self, zeta):
+        """
+        Build the moments of the self-energy via convolution.
+        """
+        cput0 = (lib.logger.process_clock(), lib.logger.perf_counter())
+        lib.logger.info(self.gw, "Building self-energy moments")
+        lib.logger.debug(self.gw, "Memory usage: %.2f GB", self._memory_usage())
+
+        q0, q1 = self.mpi_slice(self.mo_energy_g.size)
+        eta = np.zeros((q1 - q0, self.nmom_max + 1, self.nmo, self.nmo))
+        for n in range(self.nmom_max + 1):
+            zeta_prime = np.einsum("PQ,QR,RS->PS", self.Z, zeta[n], self.Z)
+            for x in range(q1 - q0):
+                Lp = np.einsum("pP,P->Pp", self.integrals.Coll, self.integrals.Coll[x])
+                eta[x, n] = np.einsum(f"Pp,Qq,PQ->pq", Lp, Lp, zeta_prime) * 2.0
+        cput1 = lib.logger.timer(self.gw, "rotating DD moments", *cput0)
+
+        moments_occ = np.zeros((self.nmom_max + 1, self.nmo, self.nmo))
+        moments_vir = np.zeros((self.nmom_max + 1, self.nmo, self.nmo))
+        moms = np.arange(self.total_nmom)
+        for n in moms:
+            fp = binom(n, moms)
+            fh = fp * (-1) ** moms
+            if np.any(self.mo_occ_g[q0:q1] > 0):
+                eo = np.power.outer(self.mo_energy_g[q0:q1][self.mo_occ_g[q0:q1] > 0], n - moms)
+                to = np.einsum(f"t,kt,ktpq->pq", fh, eo, eta[self.mo_occ_g[q0:q1] > 0])
+                moments_occ[n] += to
+            if np.any(self.mo_occ_g[q0:q1] == 0):
+                ev = np.power.outer(self.mo_energy_g[q0:q1][self.mo_occ_g[q0:q1] == 0], n - moms)
+                tv = np.einsum(f"t,ct,ctpq->pq", fp, ev, eta[self.mo_occ_g[q0:q1] == 0])
+                moments_vir[n] += tv
+        moments_occ = 0.5 * (moments_occ + moments_occ.swapaxes(1, 2))
+        moments_vir = 0.5 * (moments_vir + moments_vir.swapaxes(1, 2))
+        cput1 = lib.logger.timer(self.gw, "constructing SE moments", *cput1)
+        return moments_occ, moments_vir
```

Comment:
I will break this function up in the original `TDA` implementation so that we can inherit more of this code - ignore this comment and I will address this